### ISSUE TITLE
lz4 Windows releases no longer include a dll

### ISSUE THREAD

**Issue by tannerhelland** [lz4/lz4#787]
Through release 1.8.3, lz4 releases for Windows included pre-built shared libraries (e.g. dll/liblz4.so.1.8.3.dll).  

Starting with the 1.9.X line, prebuilt dlls are no longer present, and instead, only a static liblz4.dll.a file is shipped in the /dll folder.  It seems an unconventional change to provide a GCC-style static file here, especially when similar libraries (e.g. zstd) stick with the normal convention of shipping a prebuilt Windows dll in the /dll folder.  (Prebuilt shared libraries are incredibly convenient for Windows devs - thank you to whomever built these in the past!)

A quick scan of lz4's 1.9.X release notes doesn't explain why this changed, so I just wanted to see if this was an intentional or accidental change?

tl;dr - are there plans to include prebuilt Windows DLLs in future lz4 releases?  Similarly, is there a reason they are no longer provided (hopefully no known problems with building lz4 this way using the provided VS solutions)?

Thank you!

**Comment by Cyan4973** [comment]
The `lz4` Windows packages are created automatically by a script on Appveyor.
I don't know why it changed lately, it's probably an unintended consequence after Appveyor script modifications, which were designed at improving test coverage, but may have impacted the Windows packages creation script of source data. 
To be investigated.

**Comment by IlyaGrebnov** [comment]
@Cyan4973, are you still looking into addressing this for future releases and including official windows builds as part of release?

**Comment by Cyan4973** [comment]
Yes, at some point in the future.
On short term though, my bandwidth is limited...

**Comment by ojeda** [comment]
Given the dates between 1.8.3 and 1.9.0, perhaps it was one of the commits between 31ce8b56..0360981f to `lib/Makefile`.

**Comment by Cyan4973** [comment]
So indeed, something changed, and the name of produced shared library is now called `msys-lz4-1.dll` for some reason,
which resulted in evading the detection pattern `liblz4*`.
I'll have to find out why the change, what is the more appropriate name, and ensure it ends up in the archive.

**Comment by Cyan4973** [comment]
So what happens is that a contributor, trying to update Windows build,
decided to make many modifications related to Windows,
but many of them were hidden out of view, in their own separated file, which was then included.

When it comes to library name, the change of name happens outside of `lib/`, here : 
https://github.com/lz4/lz4/commit/b48f823f907e7763ed4178726c00453fb2af7a28#diff-45f338d2aca0e170b4531d9554b6a2a0ad301a651de3a0c237af4365dabbf2ffR17

This set the shared library name to `msys-lz4-$(LIBVER_MAJOR)`.
This decision is triggered by detection of the `MSYS%` environment.
This test comes only after detection attempts for `Windows%` and `MINGW%`,
meaning that neither were detected in the Appveyor environment.

This is surprising because I was convinced that Appveyor was providing a `MINGW64` environment.
I'm not sure about the consequences with regards to the generated binaries.
Apparently, the generated program `lz4.exe` nonetheless works correctly on Windows,
so I would expect a similar outcome for the dynamic library.

I'm also unsure about the `msys-lz4-XX` name. 
Is it following a documented common naming pattern for shared libraries under `msys` environment ? 
Or is it just a way to state that it was created under (and presumably for?) `msys` ?

Quite a few questions remain to dig in.

**Comment by Cyan4973** [comment]
In #950, the presence of the shared library in the produced package seems fixed.

Because it's produced from `msys` environment, the shared library is named `msys-lz4-1.dll`.

If you have time, you could check that the generated shared library is effectively usable :
https://ci.appveyor.com/project/YannCollet/lz4-1lndh/builds/36291113/job/60tc9vjnpotqc8cp/artifacts

At the very least, the `lz4.exe` binary produced under the same environment seems to work properly within any standard Windows shell terminal like `cmd`.

**Comment by TheDorkKnight** [comment]
@Cyan4973 I believe we're still missing the DLL's import library.

These are the two files present in `dll` as of v1.9.3:
```
dll/
 |-- liblz4.dll.a     // <- this is an AR archive full of .o object files, not useful for Windows
 '-- msys-lz4-1.dll   // <- this is a usable DLL, but there is no associated import library (eg. "msys-lz4-1.lib")
```

Clients could still load the library at runtime (with `LoadLibrary()`, `FreeLibrary()`, etc.) but it would be nice if we included an import library for implicit linking.

Additionally, this is a snippet from the current [README.md](https://github.com/lz4/lz4/blob/bdc9d3b0c10cbf7e1ff40fc6e27dad5f693a00e7/lib/dll/example/README.md#lz4-windows-binary-package) distributed with the LZ4 Windows binary package:

```
- `dll\liblz4.dll`           : The DLL of LZ4 library
- `dll\liblz4.lib`           : The import library of LZ4 library for Visual C++
```

That snippet is out of date, although it would be ideal if the released binaries returned to that naming format: it's a lot more conventional.

I suspect the build system must somehow be using the export library definition [here](https://github.com/lz4/lz4/commit/b48f823f907e7763ed4178726c00453fb2af7a28#diff-45f338d2aca0e170b4531d9554b6a2a0ad301a651de3a0c237af4365dabbf2ffR11) instead of the proper one [here](https://github.com/lz4/lz4/commit/b48f823f907e7763ed4178726c00453fb2af7a28#diff-45f338d2aca0e170b4531d9554b6a2a0ad301a651de3a0c237af4365dabbf2ffR8)
